File: data_manipulation/1_easy/actors_atleast_cooperated_3_times/actors_atleast_cooperated_3_times.scala
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.functions._

class ActorsAtleastCooperated3TimesCalculator {
  def calculate(actorDirectorDf: DataFrame): DataFrame = {
    actorDirectorDf
      .groupBy(col("actor_id"), col("director_id"))
      .agg(count(lit(1)).as("cooperation_count"))
      .filter(col("cooperation_count") >= 3)
      .select(col("actor_id"), col("director_id"))
  }
}

object ActorsAtleastCooperated3Times {
  def solve(actorDirectorDf: DataFrame): DataFrame = {
    val calculator = new ActorsAtleastCooperated3TimesCalculator()
    calculator.calculate(actorDirectorDf)
  }

  def run(spark: SparkSession): DataFrame = {
    val actorDirectorDf = spark.table("actordirector")
    solve(actorDirectorDf)
  }
}

File: data_manipulation/1_easy/actors_atleast_cooperated_3_times/actors_atleast_cooperated_3_times.py
from pyspark.sql import DataFrame, functions as F


def actors_atleast_cooperated_3_times(actordirector: DataFrame) -> DataFrame:

    
        return actordirector.groupBy(
                        F.col("actor_id"),
                        F.col("director_id")
                ).agg(
                        F.count(F.lit(1)).alias("actor_cooperated_count")
                ).filter(
                        F.col("actor_cooperated_count")>=3
                ).select(
                        F.col("actor_id"),
                        F.col("director_id")
                )
    